From f744af2d74665b7d63de75614885a8c1c75d3b51 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Thu, 29 Sep 2022 14:54:00 +0200
Subject: [PATCH] remoteproc: Add firmware format as part of sysfs interface

Add information about the supported firmware image format as
part of the sysfs interface.
This information allows the user to determine which firmware image
format to provide depending on the component in charge of loading
the remote processor firmware:
- "ELF" format if the firmware is loaded by the Linux kernel
- "TEE" format if the firmware is loaded by the TEE.

In the case of TEE format, the Linux kernel does not analyze it
but provides the image to the TEE.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I2c85234884d93f667d01ef5db7d352db9e0f133c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/268345
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 drivers/remoteproc/remoteproc_sysfs.c | 22 ++++++++++++++++++++++
 include/linux/remoteproc.h            | 13 +++++++++++++
 2 files changed, 35 insertions(+)

diff --git a/drivers/remoteproc/remoteproc_sysfs.c b/drivers/remoteproc/remoteproc_sysfs.c
index 8c7ea8922638..f356614f2975 100644
--- a/drivers/remoteproc/remoteproc_sysfs.c
+++ b/drivers/remoteproc/remoteproc_sysfs.c
@@ -174,6 +174,17 @@ static const char * const rproc_state_string[] = {
 	[RPROC_LAST]		= "invalid",
 };
 
+/*
+ * A firmware-format-to-string lookup table, for exposing a human readable
+ * supported firmware format via sysfs. Always keep in sync with enum
+ * rproc_fw_format.
+ */
+static const char * const rproc_format_string[] = {
+	[RPROC_FW_ELF]		= "ELF",
+	[RPROC_FW_TEE]		= "TEE",
+	[RPROC_FW_LAST]		= "invalid",
+};
+
 /* Expose the state of the remote processor via sysfs */
 static ssize_t state_show(struct device *dev, struct device_attribute *attr,
 			  char *buf)
@@ -235,12 +246,23 @@ static umode_t rproc_is_visible(struct kobject *kobj, struct attribute *attr,
 	return mode;
 }
 
+/* Expose the format of the remote processor binary firmware via sysfs */
+static ssize_t fw_format_show(struct device *dev, struct device_attribute *attr,
+			      char *buf)
+{
+	struct rproc *rproc = to_rproc(dev);
+	unsigned int fw_format;
+	fw_format = rproc->fw_format > RPROC_FW_LAST ? RPROC_FW_LAST : rproc->fw_format;
+	return sprintf(buf, "%s\n", rproc_format_string[fw_format]);
+}
+static DEVICE_ATTR_RO(fw_format);
 static struct attribute *rproc_attrs[] = {
 	&dev_attr_coredump.attr,
 	&dev_attr_recovery.attr,
 	&dev_attr_firmware.attr,
 	&dev_attr_state.attr,
 	&dev_attr_name.attr,
+	&dev_attr_fw_format.attr,
 	NULL
 };
 
diff --git a/include/linux/remoteproc.h b/include/linux/remoteproc.h
index b4795698d8c2..ea151949aaea 100644
--- a/include/linux/remoteproc.h
+++ b/include/linux/remoteproc.h
@@ -467,6 +467,18 @@ enum rproc_dump_mechanism {
 	RPROC_COREDUMP_INLINE,
 };
 
+/**
+ * enum rproc_fw_format - firmware image format supported
+ * @RPROC_FW_ELF:    the firmware image must be in ELF format
+ * @RPROC_FW_TEE:    the firmware image must be in format that will be treated by the TEE
+ * @RPROC_FW_LAST:   just keep this one at the end
+ */
+enum rproc_fw_format {
+	RPROC_FW_ELF,
+	RPROC_FW_TEE,
+	RPROC_FW_LAST,
+};
+
 /**
  * struct rproc_dump_segment - segment info from ELF header
  * @node:	list node related to the rproc segment list
@@ -585,6 +597,7 @@ struct rproc {
 	u16 elf_machine;
 	struct cdev cdev;
 	bool cdev_put_on_release;
+	unsigned int fw_format;
 	DECLARE_BITMAP(features, RPROC_MAX_FEATURES);
 };
 
-- 
2.39.5

